x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/queries/SparseVectorQueryBuilder.java

```diff
@@ -0,0 +1,361 @@
+        this.weightedTokensSupplier = weightedTokensSupplier;
+    }
+
+    public String getFieldName() {
+        return fieldName;
+    }
+
+    public List<WeightedToken> getQueryVectors() {
+        return queryVectors;
+    }
+
+    public boolean shouldPruneTokens() {
+        return shouldPruneTokens;
+    }
+
+    public TokenPruningConfig getTokenPruningConfig() {
+        return tokenPruningConfig;
+    }
+
+    @Override
+    protected void doWriteTo(StreamOutput out) throws IOException {
+        out.writeString(fieldName);
+        out.writeOptionalBoolean(shouldPruneTokens);
+        out.writeOptionalCollection(queryVectors);
+        out.writeOptionalString(inferenceId);
+        out.writeOptionalString(text);
+        out.writeOptionalWriteable(tokenPruningConfig);
+    }
+
+    @Override
+    protected void doXContent(XContentBuilder builder, Params params) throws IOException {
+        builder.startObject(NAME);
+        builder.field(FIELD_FIELD.getPreferredName(), fieldName);
+        if (queryVectors != null) {
+            builder.startObject(QUERY_VECTOR_FIELD.getPreferredName());
+            for (var token : queryVectors) {
+                token.toXContent(builder, params);
+            }
+            builder.endObject();
+        } else {
+            builder.field(INFERENCE_ID_FIELD.getPreferredName(), inferenceId);
+            builder.field(TEXT_FIELD.getPreferredName(), text);
+        }
+        builder.field(PRUNE_FIELD.getPreferredName(), shouldPruneTokens);
+        if (tokenPruningConfig != null) {
+            builder.field(PRUNING_CONFIG_FIELD.getPreferredName(), tokenPruningConfig);
+        }
+        boostAndQueryNameToXContent(builder);
+        builder.endObject();
+    }
+
+    @Override
+    protected Query doToQuery(SearchExecutionContext context) throws IOException {
+        if (queryVectors == null) {
+            return new MatchNoDocsQuery("Empty query vectors");
+        }
+
+        final MappedFieldType ft = context.getFieldType(fieldName);
+        if (ft == null) {
+            return new MatchNoDocsQuery("The \"" + getName() + "\" query is against a field that does not exist");
+        }
+
+        final String fieldTypeName = ft.typeName();
+        if (fieldTypeName.equals(ALLOWED_FIELD_TYPE) == false) {
+            throw new ElasticsearchParseException(
+                "field [" + fieldName + "] must be type [" + ALLOWED_FIELD_TYPE + "] but is type [" + fieldTypeName + "]"
+            );
+        }
+
+        return (this.tokenPruningConfig == null)
+            ? WeightedTokensUtils.queryBuilderWithAllTokens(queryVectors, ft, context)
+            : WeightedTokensUtils.queryBuilderWithPrunedTokens(fieldName, tokenPruningConfig, queryVectors, ft, context);
+    }
+
+    @Override
+    protected QueryBuilder doRewrite(QueryRewriteContext queryRewriteContext) {
+
+        // Note: We return a WeightedTokensQueryBuilder here instead of a SparseVectorQueryBuilder, so that this
```

Comment:
Calling out this comment - I was not able to get this to work with ccs compatibility checks enabled by returning `SparseVectorQueryBuilder` here. I think this is OK, we just need to update this to return the correct `SparseVectorQueryBuilder` when we eventually remove `WeightedTokensQueryBuilder`.

Reply:
Is the ccs compatibility failing just because this is a new query and it isn't known by the remote cluster? If so, I think it is fine?

This is a new query, why do we want to support older nodes?

Reply:
> Is the ccs compatibility failing just because this is a new query and it isn't known by the remote cluster?

Yes, that's why the ccs check is failing. 

Why I decided to go this route temporarily: while this doesn't matter for the minor version that we release it in, this allows us to pass cross cluster checks in subsequent versions. The change isn't too egregious and is easy to fix/remove when we decide we want to officially deprecate weighted_tokens queries, so it doesn't seem harmful in the short term to get that added peace of mind. 

If we decide that we _don't_ want to support this, I'd like some guidance on how we can communicate that and if there are any additional processes I should be aware of?

Reply:
> If we decide that we don't want to support this, I'd like some guidance on how we can communicate that and if there are any additional processes I should be aware of?

Its a new query, of course its not supported on older node versions. I am not sure what needs to be communicated?

Or is adding this new query breaking other CCS things for GA'd features?

Reply:
While it's true that we could decide that we don't want to support this, there are a few good reasons to consider it: 

1. When people launch Elasticsearch locally using `yarn es snapshot` for example, cross cluster compatibility checks are on by default. This means that even in 8.16, SAs and other people running demos are going to run into headaches if this fails CCS checks. I've been looking but I'm not aware of a way to fix this without only using supported queries under the hood. 
2. Even though the changes to `WeightedTokensQueryBuilder` are temporary, I think it's a bit clearer to see what it's doing right now, so it's a net positive change. 

I'm definitely open to discussion on this, I was hoping to get a little more feedback on this question.

Reply:
Resolved internally.